driver-sync/src/test/functional/com/mongodb/client/AbstractClientSideEncryptionAwsCredentialFromEnvironmentTest.java

```diff
@@ -94,4 +91,188 @@ public void testGetCredentialsFromEnvironment() {
             }
         }
     }
+    @Test
+    public void testGetCredentialsFromSupplier() {
+        assumeTrue(serverVersionAtLeast(4, 2));
+        assumeFalse(System.getenv().containsKey("AWS_ACCESS_KEY_ID"));
+        assumeTrue(isClientSideEncryptionTest());
+
+        Map<String, Map<String, Object>> kmsProviders = new HashMap<String, Map<String, Object>>() {{
+            put("aws", new HashMap<>());
+        }};
+
+        Map<String, Supplier<Map<String, Object>>> kmsProviderSupplierMap = new HashMap<String, Supplier<Map<String, Object>>>() {{
+            put("aws", () -> new HashMap<String, Object>() {{
+                put("accessKeyId", System.getProperty("org.mongodb.test.awsAccessKeyId"));
+                put("secretAccessKey", System.getProperty("org.mongodb.test.awsSecretAccessKey"));
+            }});
+        }};
+
+        try (ClientEncryption clientEncryption = createClientEncryption(ClientEncryptionSettings.builder()
+                .keyVaultNamespace("test.datakeys")
+                .kmsProviders(kmsProviders)
+                .kmsProviderSupplierMap(kmsProviderSupplierMap)
+                .keyVaultMongoClientSettings(Fixture.getMongoClientSettings())
+                .build())) {
+
+            // If this succeeds, then it means credentials have been fetched from the supplier as expected
+            BsonBinary dataKeyId = clientEncryption.createDataKey("aws", new DataKeyOptions().masterKey(
+                    BsonDocument.parse(MASTER_KEY)));
+
+            String base64DataKeyId = Base64.getEncoder().encodeToString(dataKeyId.getData());
+
+            Map<String, BsonDocument> schemaMap = new HashMap<>();
+            schemaMap.put("test.coll", getSchema(base64DataKeyId));
+            AutoEncryptionSettings autoEncryptionSettings = AutoEncryptionSettings.builder()
+                    .kmsProviders(kmsProviders)
+                    .kmsProviderSupplierMap(kmsProviderSupplierMap)
+                    .keyVaultNamespace("test.datakeys")
+                    .schemaMap(schemaMap)
+                    .build();
+            try (MongoClient client = createMongoClient(getMongoClientSettingsBuilder()
+                    .autoEncryptionSettings(autoEncryptionSettings)
+                    .build())) {
+                // If this succeeds, then it means credentials have been fetched from the supplier as expected
+                client.getDatabase("test").getCollection("coll")
+                        .insertOne(new Document("encryptedField", "encryptMe"));
+            }
+        }
+    }
+
+    @Test
+    public void shouldThrowMongoConfigurationIfSupplierThrows() {
```

Comment:
It appears that at least `shouldThrowMongoConfigurationIfSupplierThrows`, `shouldThrowMongoConfigurationIfSupplierReturnsNull` and `shouldThrowMongoConfigurationIfSupplierReturnsEmptyMap` can be refactored into a parametrized test. Maybe, even `testGetCredentialsFromSupplier` and `shouldIgnoreSupplierIfKmsProviderMapValueIsNotEmpty` could also be there.

Reply:
I combined `shouldThrowMongoConfigurationIfSupplierThrows`, `shouldThrowMongoConfigurationIfSupplierReturnsNull` and `shouldThrowMongoConfigurationIfSupplierReturnsEmptyMap` into a single parameterized test.  

I didn't do the other ones because they have different assumptions.  I thought it would be confusing to abstract the assumptions into parameters.